src/lib/result.js

```diff
@@ -1,57 +1,106 @@
 // @flow
+import { addIndex, concat, map, merge, mergeWith, reduce } from 'ramda';
 
 /**
- * Types
+ * Error messages are represented as strings
  */
 export type Errors = Array<string>;
 
-type OK = { valid: true, errors: [] };
-type Err = { valid: false, errors: Errors };
+/**
+ * A Result can be OK, which is valid, or an Err, which is invalid. An Err
+ * an array of errors, an optional type, and optional nested error items
+ */
+type OK = {| valid: true |};
+
+type Items = {
+  [string]: Result
+};
+
+type ErrType = 'array' | 'object';
+
+type Err = {|
+  valid: false,
+  errors: Errors,
+  type?: ErrType,
+  items?: Items
+|};
 
 export type Result = OK | Err;
 
 /**
- * Constructors
+ * Constructs an OK result
  */
 type _ok = () => OK;
-export const ok: _ok = () => ({ valid: true, errors: [] });
+export const ok: _ok = () => ({ valid: true });
 
-type _err = (errs?: Errors) => Err;
-export const err: _err = (errs = []) => ({ valid: false, errors: errs });
+/**
+ * Constructs an Err result
+ */
+type _err = (errors?: string | Errors, type?: ErrType, items?: Items) => Result;
+export const err: _err = (errors = [], type, items) => {
+  const result = {
+    valid: false,
+    errors: Array.isArray(errors) ? errors : [errors]
+  };
+  return items && type ? merge(result, { type, items }) : result;
+};
 
 /**
- * Helpers
+ * Is the given result OK?
  */
 type IsOK = Result => boolean;
-export const isOK: IsOK = r => r.valid === true;
+export const isOK: IsOK = ({ valid }) => valid !== undefined && valid;
```

Comment:
Is it not possible to ensure `valid` is always a boolean with Flow, so we don't need to do these null checks?

Reply:
I think the issue might be that Flow isn't smart enough to follow the destructuring; to be honest though I could just undo this line because it was fine before 😛